README.md

MangoTest
=========

R package containing answers to the following Mango Solutions recruitment questions. See `docs/index.html` for package documentation.

Questions
---------

### Data analysis

-   Read in the csv containing NBA free throws.
    -   See `data-raw/read_in_free_throws.R`.
-   Visualise some aspect of the data you find interesting, e.g., the average number of free throws per period for the regular season and the playoffs.
    -   See `vignettes/data_analysis.Rmd` and `vignettes/drafts/data_analysis/data_analysis.html`.

### Programming

#### Function

-   Write a function that will draw random numbers from a given distribution. The function should take one argument for the number of samples and a second argument which specifies the distribution (Normal, Poisson or Binomial). The function should be able to handle additional parameters depending on the distribution chosen, e.g., a `lambda` for Poisson samples.
    -   See `R/sample_dist.R` for the function, `tests/testhat/sample_dist.R` for function tests, and `docs/reference/sample_dist.html` for examples.

#### Object-oriented programming

-   Use the class system of your choice to create a summary method for the output of your function.
    -   See `R/summary.sample_dist`, `tests/testhat/summary.sample_dist` for function tests, and `docs/reference/summary.sample_dist.html` for examples.

#### Package

-   Build a package containing the functions above. Use any tools as necessary.
    -   See repo.

Installation
------------

Install the package using devtools with the following code.

``` r
# install.packages("devtools")
devtools::install_github("seabbs/MangoTest")
```

Development Environment
-----------------------

This package has been developed in docker based on the`rocker/tidyverse` image, to access the development environment enter the following at the command line in the package directory (with an active docker daemon running),

``` bash
docker docker . -t MangoTest
docker run -d -p 8787:8787 -e USER=MangoTest -e PASSWORD=MangoTest --name MangoTest MangoTest
```

The rstudio client can be accessed on port 8787 at localhost (or your machines ip). The default username is MangoTest and the default password is MangoTest.

If a package rebuild is required this can be achieved using `make`, as follows;

``` bash
make
```

Notes
-----

The travis and appveyor set up is for information only and has not been linked to their online services. Badges outlining the build status, along with code coverage and activity status of the repository would be added to the top of this package when made public.


NEWS.md
# MangoTest 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.
* Added `sample_dist` function. 
* Added `summary` method for class "sample_dist".
* Added unit tests.
* Added pkgdown site.
* Added NBA data analysis.


tests/testthat.R
library(testthat)
library(MangoTest)

test_check("MangoTest")


tests/testthat/test-sample_dist.R
context("sample_dist")


# Dummy function
dist <- function(n = NULL) {return(n)}

test_that("sample_dist functions correctly with a dummy function", {
  result <- rpois(1,lambda = 10)

  test <- sample_dist(result, dist = dist)
  class(test) <- "integer"
expect_equal(result, test)
  })

test_that("sample_dist errors when a number of samples is not supplied", {
  expect_error(sample_dist(dist = rnorm))
})


test_that("sample_dist errors when the number of samples is supplied as a character", {
  expect_error(sample_dist(n = "1", dist = rnorm))
})

test_that("sample_dist errors when the number of samples is not an integer", {
  expect_error(sample_dist(n = 1.2, dist = rnorm))
})

test_that("sample_dist errors when the number of samples is a vector", {
  expect_error(sample_dist(n = c(1,2), dist = rnorm))
})

test_that("sample_dist errors when a function is not passed as the distribution", {
  expect_error(sample_dist(n = 1, dist = 2))
  expect_error(sample_dist(n = 1, dist = "2"))
  expect_error(sample_dist(n = 1, dist = data.frame(x = 1)))
})


tests/testthat/test-summary.sample_dist.R
context("summary.sample_dist")



test <- sample_dist(1000, dist = rnorm)


result <- summary(test)

test_summary_measure <- function(test, result, sum_fn, fn_name, ...) {
  test_that(paste0("summary.sample_dist calculates the ", fn_name,  " correctly"), {
    expect_equal(as.numeric(result), as.numeric(sum_fn(test, ...)))
  })
}

test_summary_measure(test, result["Mean"], sum_fn = mean, fn_name = "mean")
test_summary_measure(test, result["Median"], sum_fn = median, fn_name = "median")
test_summary_measure(test, result["2.5%"], sum_fn = quantile, fn_name = "quantile", probs = 0.025)
test_summary_measure(test, result["Maximum"], fn_name = "max", sum_fn =  max)
test_summary_measure(test, result["Minimum"], fn_name = "min", sum_fn =  min)


R/free_throws.R
#' NBA Free Throws Data
#'
#' A dataframe of NBA free throw data as supplied by Mango Solutions
#' @format A data frame with 618,019 rows and 11 variables.
#' \describe{
#'   \item{end_result}{Character, final score of the game.}
#'   \item{game}{Character, identifies the two teams playing.}
#'   \item{game_id}{Numeric, game id.}
#'   \item{period}{Numeric, which period was the play made in.}
#'   \item{play}{Character, outlining the play.}
#'   \item{player}{Character, the name of the player.}
#'   \item{playoffs}{Character, type of game.}
#'   \item{score}{Character, identifying the final score.}
#'   \item{season}{Character, identifying the season the game was played in.}
#'   \item{shot_made}{Binary, was a shot made.}
#'   \item{time}{Time at which a play occurred.}
#' }
"free_throws"


R/sample_dist.R

#' Sample from a Probability Distribution
#'
#' @description Function brief: Write a function that will draw random numbers
#'  from a given distribution. The function should take one argument for the
#'  number of samples and a second argument which specifies the distribution
#'  (Normal, Poisson or Binomial). The function should be able to handle
#'  additional parameters depending on the distribution chosen, e.g., a `lambda`
#'   for Poisson samples.
#' @param n Numeric (or integer), the number of samples to take.
#' @param dist A function which samples from a probability distribution. The first
#' arguement must be the number of samples, additional arguements can be specified with
#' \code{...}. Possible options are \code{rnorm} (normal), \code{rpois} (Poisson), and
#' \code{rbinom} (Binomial).
#' @param ... Additional arguements to pass to the specified \code{dist} function.
#' @return A vector of samples from the specified distribution.
#' @importFrom stats quantile
#' @export
#'
#' @examples
#'
#' ## Sample a single time from the normal distribution
#' sample_dist(1, dist = rnorm, mean = 0, sd = 1)
#'
#' ## Sample multiple times from the poisson distribution
#' sample_dist(100, dist = rpois, lambda = 2)
#'
sample_dist <- function(n = NULL, dist = NULL, ...) {
  if (is.null(n)) {
    stop("The number of samples must be specified, using n.")
  }

  if (!is.numeric(n)) {
    stop("The number of samples (n) must be numeric.")
  }

  if (length(n) > 1) {
    stop("The number of samples must be a single number.")
  }

  if (!(n %% 1 == 0)) {
    stop("The number of samples must be an integer")
  }

  if (!is.function(dist)) {
    stop("dist must be a function to sample from a probability distribution.
         An example for the normal distribution is rnorm")
  }

  samples <- dist(n, ...)

  class(samples) <- c("sample_dist", class(samples))

  return(samples)
}


R/summary.sample_dist.R

#' Summarising a sample of a probability distribution.
#'
#' @description \code{summary} method for class "sample_dist".
#' @param object A numeric vector of class "sample_dist".
#' @param plot Logical, defaults to \code{TRUE}. Produces a summary plot.
#' @param ... Pass additional arguements to summary function.
#' @seealso summary
#' @aliases summary
#' @method summary sample_dist
#' @return Prints a summary table to the terminal and a summary plot.
#' @export
#' @import ggplot2
#' @examples
#' ## Sample from a normal distribution
#' x <- sample_dist(100, dist = rnorm)
#'
#' ## Create a summary
#' summary(x)
summary.sample_dist <- function(object = NULL, plot = TRUE, ...) {
  if (is.null(object)) {
    stop("A vector of samples from a probability distribution is required")
  }

  if (!is.numeric(object)) {
    stop("The vector of probability samples must be numeric")
  }

  n <- length(object)
  min <- min(object)
  max <- max(object)
  mean <- mean(object)
  median <- median(object)
  lll <- quantile(object, 0.025)
  ll <- quantile(object, 0.25)
  hh <- quantile(object, 0.75)
  hhh <- quantile(object, 0.975)

  sum_list <- c(Samples = n,
                Minimum = min,
                lll,
                ll,
                Median = median,
                Mean = mean,
                hh,
                hhh,
                Maximum = max
                )

  rest_sum_list <- sum_list[c("2.5%", "25%", "Mean", "Median", "75%", "97.5%")]

  if (plot) {
    samples <- NULL
    value <- NULL
    Measures <- NULL
    quantile <- NULL

    plot <- ggplot(data.frame(samples = object),
                   aes(x = samples, fill = 1)) +
      geom_vline(data = data.frame(Measures = names(rest_sum_list),
                                   value = rest_sum_list),
                 aes(xintercept = value,
                     linetype = Measures)) +
      geom_density(alpha = 0.5, show.legend = c(fill = FALSE)) +
      geom_rug() +
      theme_minimal() +
      theme(legend.position = "bottom") +
      guides(linetype = guide_legend(title = "Summary measures")) +
      labs(title = "Distribution of Probability Samples",
           caption = paste0("Showing distribution of ", n, " samples"),
           y = "Density",
           x = "Samples")

    print(plot)
  }

  return(sum_list)
}
